lib/compress/index.js

```diff
@@ -5603,15 +5603,35 @@ def_optimize(AST_Call, function(self, compressor) {
             if (has_flag(arg, UNUSED)) continue;
             if (!safe_to_inject
                 || block_scoped.has(arg.name)
-                || identifier_atom.has(arg.name)
-                || scope.conflicting_def(arg.name)) {
+                || identifier_atom.has(arg.name)) {
                 return false;
             }
+            if (scope.conflicting_def(arg.name)) {
+                rename_arg(arg);
+            }
             if (in_loop) in_loop.push(arg.definition());
         }
         return true;
     }
 
+    function rename_arg(arg) {
+        const newArgName = fn.create_symbol(AST_SymbolFunarg, {
+            source: fn,
+            scope: scope,
+            tentative_name: "argument_" + fn.argnames.length,
+        });
+
+        const cb = (node) => {
+            if (node instanceof AST_SymbolRef && node.thedef.name === arg.name) {
+                node.thedef.name = newArgName;
+                return node;
+            } else {
+                return walk(node, cb);
+            }
+        };
```

Comment:
In lib/scope.js is an easy example of how to rename a variable. The definition object (obtained through `node.thedef` or `node.definition()`) can help us iterate on references and definitions alike.

It also deal with shadowing (in case `fn` here contains another scope which redefines a variable of the same name)

```
def.name = name
def.orig.forEach(function(sym) {      
    sym.name = name;                  
});                                   
def.references.forEach(function(sym) {
    sym.name = name;                  
});                                   
```

Reply:
Otherwise this looks great, and the tests are sound.

Reply:
I edited the renaming code. However, I don't know why the original function doesn't seem to be updated. The last test should fail as I have written it. The inlined argument should be named `argument_1`.

Reply:
Maybe the cause is the fact that the "defaults" option is enabled, causing more transforms to occur.

If you remove defaults and add reduce_vars, inline and unused, it should showcase what you did a bit better. Even if the code in the "expect" section isn't truly optimally minified, which isn't so important for a test.

Reply:
I should have a deeper review of this PR later today. My cursory look says it's good to merge, but I'll check it out better after work.

Reply:
After playing more with this, I actually think this is the correct behavior! I wanted to rename the variables of the pure function so they can be inlined by the compressor (like the second test). The compressor correctly substitutes the new name we created with the argument passed in!